core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleCall.java

```diff
@@ -160,6 +160,11 @@ protected void onMatch() {
         return;
       }
 
+      if (isRuleExcluded()) {
+        LOGGER.debug("Rule [{}] not fired due to exclusion hint", getRule());
```

Comment:
Should this be evaluated before the rule is added to the match queue?

Reply:
It's hard to do that, because the operands can be matched from children to parent (ASENDING), for current implementation, i only want to match the root operand of the final match.